src/core/weights.js

```diff
@@ -48,19 +52,58 @@ export function copy(w: Weights): Weights {
   };
 }
 
-/**
- * Merge multiple Weights together.
+/** Merge multiple Weights together.
+ *
+ * The resultant Weights will have every weight specified by each of the input
+ * weights.
  *
- * The resultant Weights will have every weight specified by each of the
- * input weights. If there are any overlaps (i.e. the same address is present
- * in two or more of the input weights), an error will be thrown. In the future,
- * we will likely modify this function to add a resolver that determines how to
- * combine multiple overlapping weights.
+ * When there are overlaps (i.e. the same address is present in two or more
+ * input weights), the appropriate resolver will be invoked to resolve the
+ * conflict. The resolver will receive, as its first argument, the result of
+ * resolving the last conflict (or the first weight if there have been no
+ * conflicts), and its second argument will be the latest conflicting weight.
+ * Whatever it returns will be used as the weight (modulo future conflicts).
  */
-export function merge(ws: $ReadOnlyArray<Weights>): Weights {
-  const nodeWeights = MapUtil.merge(ws.map((x) => x.nodeWeights));
-  const edgeWeights = MapUtil.merge(ws.map((x) => x.edgeWeights));
-  return {nodeWeights, edgeWeights};
+export function merge(
+  ws: $ReadOnlyArray<Weights>,
+  nodeResolver: ?NodeOperator,
+  edgeResolver: ?EdgeOperator
+): Weights {
+  const weights: Weights = empty();
+  const {nodeWeights, edgeWeights} = weights;
+  if (nodeResolver == null) {
+    nodeResolver = (_unused_a, _unused_b) => {
+      throw new Error(
+        "node weight conflict detected, but no resolver specified"
+      );
+    };
+  }
+  if (edgeResolver == null) {
+    edgeResolver = (_unused_a, _unused_b) => {
+      throw new Error(
+        "edge weight conflict detected, but no resolver specified"
+      );
+    };
+  }
+  for (const w of ws) {
+    for (const [addr, val] of w.nodeWeights.entries()) {
+      const existing = nodeWeights.get(addr);
+      if (existing == null) {
+        nodeWeights.set(addr, val);
+      } else {
+        nodeWeights.set(addr, nodeResolver(existing, val));
```

Comment:
Suggest a try-catch around the resolver, to add the `NodeAddress` to the error message.
In case of a bug where we're expecting disjoint weights, this seems like important debug info :]

Reply:
Good idea. I updated it, and checked the behavior with some new tests.